### PR TITLE
Reduce usage of fmt in serialization and error reporting

### PR DESCRIPTION
**Pull request by SimonSapin**
`format!` and `write!` create a somewhat-heavyweight `Formatting` struct and use dynamic dispatch to call into impls of `Dispaly` and related traits. The former also allocates an intermediate string that is sometimes unnecessary.

I started looking into this from https://bugzilla.mozilla.org/show_bug.cgi?id=1355599, but I expect the impact there will be small to insignificant. It might be a slightly less so on parsing (error reporting).

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/18355)

### CODE REVIEW

File: ports/geckolib/error_reporter.rs [hunk 1]
```diff
@@ -53,137 +53,14 @@ impl<'a> ErrorString<'a> {
...
             }
-            escaped.push(ch);
         }
     }
-
-    escaped
-}
-
-// This is identical to the behaviour of cssparser::CssStringWriter, except that
-// the characters between 0x7F and 0xA0 as numerically escaped as well.
```
Comment (emilio) [inline ports/geckolib/error_reporter.rs]: Did we agree to change the test then? Nice to see this going away.

File: components/style_traits/values.rs [hunk 2]
```diff
@@ -533,14 +551,3 @@ pub mod specified {
         }
     }
 }
-
-
-/// Wrap CSS types for serialization with `write!` or `format!` macros.
-/// Used by ToCss of SpecifiedOperation.
-pub struct Css<T>(pub T);
```
Comment (emilio) [inline components/style_traits/values.rs]: Nice, one more hack going away.

File: tests/unit/stylo/specified_values.rs [hunk 3]
```diff
@@ -5,6 +5,7 @@
 use style;
 
 #[cfg(all(test, target_pointer_width = "64"))]
+#[test]
```
Comment (emilio) [inline tests/unit/stylo/specified_values.rs]: lol

**Review by emilio — APPROVED** [review]

**Comment by SimonSapin** [comment]
@dbaron, in https://bugzilla.mozilla.org/show_bug.cgi?id=229827#c7 you suggested escaping U+0080 to U+009F in CSS syntax in error messages, whereas https://drafts.csswg.org/cssom/#serialize-an-identifier specifies to leave all non-ASCII code points as-is. Error messages are not subject to interop concerns, but it’d be nicer to only keep one code path for CSS serialization. Additionally, since Rust/cssparser are based on UTF-8, it is easier (and possibly more efficient) to work on UTF-8 bytes rather than code points, which in turn is easier if all non-ASCII bytes are treated the same.

So, together with this PR, I’d like to land a patch that makes some expected results of https://dxr.mozilla.org/mozilla-central/source/layout/style/test/test_parser_diagnostics_unprintables.html conditional on `isStylo`, where all non-ASCII code points are left unescaped. What do you think?

**Comment by SimonSapin** [comment]
```diff
diff --git a/layout/style/test/test_parser_diagnostics_unprintables.html b/layout/style/test/test_parser_diagnostics_unprintables.html
index f478f7916764..47fac9910b6f 100644
--- a/layout/style/test/test_parser_diagnostics_unprintables.html
+++ b/layout/style/test/test_parser_diagnostics_unprintables.html
@@ -127,26 +127,26 @@ const substitutions = [
                                          s: "\\1c \\1d \\1e \\1f  " },
 
   // U+007F (DELETE) and U+0080 - U+009F (C1 controls)
-  { t: "\\\x7f\\\x80\\\x81\\\x82",       i: "\\7f \\80 \\81 \\82 ",
-                                         s: "\\7f \\80 \\81 \\82 " },
-  { t: "\\\x83\\\x84\\\x85\\\x86",       i: "\\83 \\84 \\85 \\86 ",
-                                         s: "\\83 \\84 \\85 \\86 " },
-  { t: "\\\x87\\\x88\\\x89\\\x8A",       i: "\\87 \\88 \\89 \\8a ",
-                                         s: "\\87 \\88 \\89 \\8a " },
-  { t: "\\\x8B\\\x8C\\\x8D\\\x8E",       i: "\\8b \\8c \\8d \\8e ",
-                                         s: "\\8b \\8c \\8d \\8e " },
-  { t: "\\\x8F\\\x90\\\x91\\\x92",       i: "\\8f \\90 \\91 \\92 ",
-                                         s: "\\8f \\90 \\91 \\92 " },
-  { t: "\\\x93\\\x94\\\x95\\\x96",       i: "\\93 \\94 \\95 \\96 ",
-                                         s: "\\93 \\94 \\95 \\96 " },
-  { t: "\\\x97\\\x98\\\x99\\\x9A",       i: "\\97 \\98 \\99 \\9a ",
-                                         s: "\\97 \\98 \\99 \\9a " },
-  { t: "\\\x9B\\\x9C\\\x9D\\\x9E\\\x9F", i: "\\9b \\9c \\9d \\9e \\9f ",
-                                         s: "\\9b \\9c \\9d \\9e \\9f " },
+  { t: "\\\x7f\\\x80\\\x81\\\x82",       i: isStylo ? "\\7f \x80\x81\x82" : "\\7f \\80 \\81 \\82 ",
+                                         s: isStylo ? "\\7f \x80\x81\x82" : "\\7f \\80 \\81 \\82 " },
+  { t: "\\\x83\\\x84\\\x85\\\x86",       i: isStylo ? "\x83\x84\x85\x86" : "\\83 \\84 \\85 \\86 ",
+                                         s: isStylo ? "\x83\x84\x85\x86" : "\\83 \\84 \\85 \\86 " },
+  { t: "\\\x87\\\x88\\\x89\\\x8A",       i: isStylo ? "\x87\x88\x89\x8A" : "\\87 \\88 \\89 \\8a ",
+                                         s: isStylo ? "\x87\x88\x89\x8A" : "\\87 \\88 \\89 \\8a " },
+  { t: "\\\x8B\\\x8C\\\x8D\\\x8E",       i: isStylo ? "\x8B\x8C\x8D\x8E" : "\\8b \\8c \\8d \\8e ",
+                                         s: isStylo ? "\x8B\x8C\x8D\x8E" : "\\8b \\8c \\8d \\8e " },
+  { t: "\\\x8F\\\x90\\\x91\\\x92",       i: isStylo ? "\x8F\x90\x91\x92" : "\\8f \\90 \\91 \\92 ",
+                                         s: isStylo ? "\x8F\x90\x91\x92" : "\\8f \\90 \\91 \\92 " },
+  { t: "\\\x93\\\x94\\\x95\\\x96",       i: isStylo ? "\x93\x94\x95\x96" : "\\93 \\94 \\95 \\96 ",
+                                         s: isStylo ? "\x93\x94\x95\x96" : "\\93 \\94 \\95 \\96 " },
+  { t: "\\\x97\\\x98\\\x99\\\x9A",       i: isStylo ? "\x97\x98\x99\x9A" : "\\97 \\98 \\99 \\9a ",
+                                         s: isStylo ? "\x97\x98\x99\x9A" : "\\97 \\98 \\99 \\9a " },
+  { t: "\\\x9B\\\x9C\\\x9D\\\x9E\\\x9F", i: isStylo ? "\x9B\x9C\x9D\x9E\x9F" : "\\9b \\9c \\9d \\9e \\9f ",
+                                         s: isStylo ? "\x9B\x9C\x9D\x9E\x9F" : "\\9b \\9c \\9d \\9e \\9f " },
 
   // CSS doesn't bother with the full Unicode rules for identifiers,
   // instead declaring that any code point greater than or equal to
-  // U+00A0 is a valid identifier character.  Test a small handful
+  // U+0080 is a valid identifier character.  Test a small handful
   // of both basic and astral plane characters.
 
   // Arabic (caution to editors: there is a possibly-invisible U+200E
```

**Comment by dbaron** [comment]
I'm ok with no longer escaping U+0080 to U+009F, although it might be better to make the change in both Stylo and Gecko.

**Comment by SimonSapin** [comment]
```
  ▶ Unexpected subtest result in /_mozilla/mozilla/calc.html:
  │ FAIL [expected PASS] calc(1px + 1pt + 1pc + 1in + 1cm + 1mm)
  │   → assert_equals: expected "calc(155.91666666666666px)" but got "calc(155.916666666667px)"
```

I’ve modified the test so that the result is an integer number of pixels (in fact a multiple of 96px), to make it independent of float serialization details. r? @emilio

**Comment by SimonSapin** [comment]
@bors-servo p=1

(I need to land https://reviewboard.mozilla.org/r/176034/ around the same time)

**Comment by SimonSapin** [comment]
Updated WPT manifest.

@bors-servo r=emilio

**Pull request merged**